cni/network/invoker_cns.go

```diff
@@ -66,82 +65,101 @@ func (invoker *CNSIPAMInvoker) Add( //nolint don't consider unnamedResult
 	log.Printf(podInfo.PodName)
 	orchestratorContext, err := json.Marshal(podInfo)
 	if err != nil {
-		return nil, nil, fmt.Errorf("Failed to unmarshal orchestrator context during add: %w", err)
+		return IPAMAddResult{}, fmt.Errorf("Failed to unmarshal orchestrator context during add: %w", err)
 	}
 
-	if args == nil {
-		return nil, nil, errEmptyCNIArgs
+	if opt.args == nil {
+		return IPAMAddResult{}, errEmptyCNIArgs
 	}
 
-	ipconfig := cns.IPConfigRequest{
-		OrchestratorContext: orchestratorContext,
-		PodInterfaceID:      GetEndpointID(args),
-		InfraContainerID:    args.ContainerID,
-	}
+	ret := IPAMAddResult{subnetPrefix: &net.IPNet{}}
 
-	log.Printf("Requesting IP for pod %+v using ipconfig %+v", podInfo, ipconfig)
-	response, err := invoker.cnsClient.RequestIPAddress(context.TODO(), ipconfig)
-	if err != nil {
-		log.Printf("Failed to get IP address from CNS with error %v, response: %v", err, response)
-		return nil, nil, err
-	}
-
-	info := IPv4ResultInfo{
-		podIPAddress:       response.PodIpInfo.PodIPConfig.IPAddress,
-		ncSubnetPrefix:     response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
-		ncPrimaryIP:        response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
-		ncGatewayIPAddress: response.PodIpInfo.NetworkContainerPrimaryIPConfig.GatewayIPAddress,
-		hostSubnet:         response.PodIpInfo.HostPrimaryIPInfo.Subnet,
-		hostPrimaryIP:      response.PodIpInfo.HostPrimaryIPInfo.PrimaryIP,
-		hostGateway:        response.PodIpInfo.HostPrimaryIPInfo.Gateway,
-	}
+	if opt.nwCfg.MultiTenancy {
+		// Temporary if block to determining whether we disable SNAT on host (for multi-tenant scenario only)
+		if ret.enableSnatForDNS, opt.nwCfg.EnableSnatOnHost, err = invoker.multitenancyClient.DetermineSnatFeatureOnHost(
```

Comment:
can it be in network.go itself? why it should be in ipam invoker function?

Reply:
I thought we had a discussion on this, you said we should move this here